src/triage/component/catwalk/predictors.py

```diff
@@ -153,80 +146,43 @@ def _write_to_db(
-        else:
-            logging.info(
-                "as_of_date not found as part of matrix index, using "
-                "matrix metadata end_time as as_of_date"
-            )
-            temp_df = pandas.DataFrame({"score": predictions})
-            rankings_abs = temp_df["score"].rank(method="dense", ascending=False)
-            rankings_pct = temp_df["score"].rank(
-                method="dense", ascending=False, pct=True
-            )
-            for entity_id, score, label, rank_abs, rank_pct in zip(
-                matrix_store.index,
-                predictions,
-                labels,
-                rankings_abs,
-                rankings_pct,
+        with tempfile.TemporaryFile(mode="w+") as f:
+            writer = csv.writer(f, quoting=csv.QUOTE_MINIMAL)
+            for index, score, label in zip(
+                matrix_store.index, predictions, labels
             ):
-                db_objects.append(
-                    Prediction_obj(
-                        model_id=int(model_id),
-                        entity_id=int(entity_id),
-                        as_of_date=matrix_store.metadata["end_time"],
-                        score=round(float(score), 10),
-                        label_value=int(label) if not math.isnan(label) else None,
-                        rank_abs=int(rank_abs),
-                        rank_pct=round(float(rank_pct), 10),
-                        matrix_uuid=matrix_store.uuid,
-                        test_label_timespan=test_label_timespan,
-                        **misc_db_parameters
-                    )
+                entity_id, as_of_date = index
+                prediction = Prediction_obj(
+                    model_id=int(model_id),
+                    entity_id=int(entity_id),
+                    as_of_date=as_of_date,
+                    score=float(score),
+                    label_value=int(label) if not math.isnan(label) else None,
+                    matrix_uuid=matrix_store.uuid,
+                    test_label_timespan=test_label_timespan,
+                    **misc_db_parameters
                 )
-
-            try:
-                session = self.sessionmaker()
-                session.bulk_save_objects(db_objects)
-                session.commit()
-            finally:
-                session.close()
+                writer.writerow(
+                    [
+                        prediction.model_id,
+                        prediction.entity_id,
+                        prediction.as_of_date,
+                        prediction.score,
+                        prediction.label_value,
+                        prediction.rank_abs,
+                        prediction.rank_pct,
+                        prediction.matrix_uuid,
+                        prediction.test_label_timespan,
+                    ]
+                )
+            f.seek(0)
+            postgres_copy.copy_from(f, Prediction_obj, self.db_engine, format="csv")
```

Comment:
working on supporting exactly this with `ohio.PipeTextIO`, so you needn't bother with managing/seeking this temporary file :smile_cat:

Reply:
Interested to see it! Since you mentioned this I forgot that I had created a util within Triage for wrapping this part of it that I had just never applied to the Predictor.

https://github.com/dssg/triage/blob/master/src/triage/component/catwalk/utils.py#L202

But I can check out PipeTextIO

Reply:
Cool. Yeah, it's "done"… though, I ran into an oddity with a (different) copy utility, so I would still call it "alpha" status :smile: